rvv-intrinsic-rfc.md

```diff
@@ -602,14 +600,19 @@ This section lists all utility functions to help users program in V intrinsics e
 
 ### Vector Initialization<a name="vector-init"></a>
 
-These utility functions are used to initialize vector values. They could be used in masking intrinsics with *zero in output* and *don't care in output* semantics.
+These utility functions are used to initialize vector values. They could be used in masking intrinsics with *don't care in output* semantics.
 
 ```
 Example:
 
 vint8m1_t vundefined_i8m1()
 ```
 
+Note: Any operation with `vundefined_* ()` are undefined and unpredictable,
+the only recommended usage is used as maskedoff operand. e.g.
+`vxor(vundefined (), vundefined ())` will not guarantee result vector with
```

Comment:
My example case was:

vec a = vundefined();
vxor(a, a);

So vundefined() was only called once but constant folding inside the compiler propagated undef to both inputs. The suggestion to use (freeze undef) still wouldn't have made vxor(vundefined (), vundefined ()) produce 0 as each input would be a unique freeze and therefore not equal.

Reply:
How about
```
vec a;
vxor(a, a);
```
If the above result is 0, I think users will treat the following code also get 0.
```
vec a = vundefined();
vxor(a, a);
```

Reply:
I wouldn't expect  

```
vec a = vundefined();
vxor(a, a);
```

be equivalent to

```
vxor(vundefined(), vundefined());
```

because I understand `vundefined()` as `vrandombits()` as a non-pure function (and no defined semantics is given to the bits returned, they can be the same, or different every time, etc). Under these semantics `freeze undef` in LLVM looks to me it should work, shouldn't it?

Reply:
InstCombine folds (freeze undef) to 0 so we lose track of the fact that it was undefined. This results in an instruction to create a 0 vector. Which in a lot of places is probably good to prevent false dependencies on an out of order core. But for the purpose of using it to indicate an undefined input for a masked operation it doens't work.

Reply:
@Hsiangkai has try to implement with `freeze undef` but it seems result worse code gen when used in maskedoff operand...

Testcase:
```
#include <riscv_vector.h>
void foo(int32_t *a, int32_t *b, int len) {
  vbool32_t mask = vmset_m_b32();
  vint32m1_t va = vle32_v_i32m1_m_vl (mask, vundefined_i32m1 (), a, len);
  vse32_v_i32m1_m_vl (mask, b, va, len);
}
```
Without `freeze undef`:
```
foo:                                    # @foo
# %bb.0:                                # %entry
        csrr    a3, vl
        vsetvli a3, a3, e8,mf4,tu,mu
        vmset.m v0
        vsetvli a2, a2, e32,m1,tu,mu
        vle32.v v1, (a0), v0.t
        vse32.v v1, (a1), v0.t
        ret
```
With `freeze undef`:
```
foo:                                    # @foo
# %bb.0:                                # %entry
        csrr    a3, vl
        vsetvli a3, a3, e8,mf4,tu,mu
        vmset.m v0
        vsetvli a3, zero, e32,m1,tu,mu
        vmv.v.i v1, 0                                        # <----------------------------------- Extra 0 init instruction
        vsetvli a2, a2, e32,m1,tu,mu
        vle32.v v1, (a0), v0.t
        vse32.v v1, (a1), v0.t
        ret
```

Reply:
Oh I see, this is disappointing. I wasn't expecting that fold.

ISTR for AVX-512 clang already emits a zero vector for the equivalent undefined intrinsic. I understand doing that would incur on the same issue of creating a vector of zeroes.

LLVM-specifically: perhaps we can have a `@llvm.rvv.vundefined.<ty>()` that we select as `IMPLICIT_DEF` in SelectionDAG? Now I'm wondering if we want it to have side-effects or not as SelectionDAG will CSE it if it ends being mapped to a node without chain (and that would introduce some false dependences that @topperc mentioned). I wonder if we can give it side-effects but then (re-)chain it to the entry token of the BB. Could that work @topperc?

Reply:
I think if its chained to the entry node, then it would then become subject to CSE again. SelectionDAG doesn't exclude chain nodes from CSE. It only excludes a couple specific opcodes and nodes with with Glue.

We could keep it without side effects and replace it with undef if its used in a masked intrinsic? That would allow those to select IMPLICIT_DEF. And other usages can just select to 0? We also need to select the vtype to "mask agnostic" right? Otherwise we have a false dependency.